### PR TITLE
Allow dataloader to accept a custom memory pinning function

### PR DESCRIPTION
**Pull request by mcarilli**
Renewed attempt at https://github.com/pytorch/pytorch/pull/14171

From the original PR:
> Currently, the pin_memory_batch function in the dataloader will return a batch comprised of any unrecognized type without pinning the data, because it doesn't know how.
>
>This behavior was preventing us from overlapping data prefetching in Mask-RCNN, whose custom collate_fn returns a custom batch type. 

The old PR allowed the user to implement batch pinning for custom batch and data types by passing a custom pin function to the dataloader.  @slayton58 suggested a cleaner approach:  allow the user to define a `pin_memory` method on their custom types, and have `pin_memory_batch` [check for the presence of that method](https://github.com/pytorch/pytorch/pull/16743/files#diff-9f154cbd884fe654066b1621fad654f3R56) in the incoming batch as a fallback.  I've updated the test and docstrings accordingly.

The old PR was merged but then reverted due to weird cuda OOM errors on windows that may or may not have been related.  I have no idea why my changes would cause such errors (then or now) but it's something to keep an eye out for.

@fmassa and @yf225 who were my POCs on the old PR.

### CODE REVIEW

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -42,7 +42,9 @@ class DataLoader(object):
             (default: ``0``)
         collate_fn (callable, optional): merges a list of samples to form a mini-batch.
         pin_memory (bool, optional): If ``True``, the data loader will copy tensors
-            into CUDA pinned memory before returning them.
+            into CUDA pinned memory before returning them.  If your data elements
+            are a custom type, or your ``collate_fn`` returns a batch that is a custom type
+            see the Warning below.
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: nit: probably should use lowercase `warning`

File: torch/utils/data/dataloader.py [hunk 2]
```diff
@@ -65,6 +67,15 @@ class DataLoader(object):
...
     .. warning:: If ``spawn`` start method is used, :attr:`worker_init_fn` cannot be an
                  unpicklable object, e.g., a lambda function.
+
+    .. warning:: The default memory pinning logic only recognizes Tensors and maps and iterables
+                 containg Tensors.  By default, if the pinning logic sees a batch that is a custom type,
+                 (which will occur if you have a ``collate_fn`` that returns a custom batch type),
+                 or if each element of your batch is a custom type, the pinning logic will not
+                 recognize them, and it will return that batch (or those elements)
+                 without pinning the memory.  To enable memory pinning for custom batch or data types,
+                 define a pin_memory method on your custom type(s).  See ``SimpleCustomBatch`` and
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: Users don't have direct access to these files. I would prefer writing them in an `Example::` code block below.

**Review by ssnl — COMMENTED** [review]
please fix the two nits. looks reasonable otherwise

**Review by fmassa — COMMENTED** [review]
This looks good to me, thanks!

One question I have: isn't it more general to allow passing a custom `pin_memory` function?
While I'm ok with this PR, the assumption about `pin_memory` is a bit hidden to me, and I'd think it would make sense to expose it to the user.

Was the reason for replacing it with the `pin_memory` attribute to circumvent the potential Windows timeout?

File: torch/utils/data/dataloader.py (hunk 2 above)
Comment (mcarilli) [reply torch/utils/data/dataloader.py]: Ok, a minimal-but-complete in-place example doesn't actually look too bad there imo.  Let me know what you think.

**Comment by fmassa** [comment]
I'm ok with this new approach.

cc @gchanan to see if he has any preferences.

**Review by fmassa — APPROVED** [review]

**Review by ezyang — APPROVED** [review]
